superset-frontend/src/explore/components/controls/AnnotationLayerControl/AnnotationLayer.jsx

```diff
@@ -279,17 +292,26 @@ class AnnotationLayer extends React.PureComponent {
     });
   }
 
+  /*
+  To ensure re-rendering of the AsyncSelect annotation source component we
+  set optionsFetch here.
+  */
   handleAnnotationSourceType(sourceType) {
     const { sourceType: prevSourceType } = this.state;
 
     if (prevSourceType !== sourceType) {
-      this.setState({ sourceType, value: null, isLoadingOptions: true });
+      this.setState({
+        sourceType,
+        value: null,
+        valueOptions: {},
+        optionsFetch: this.asyncFetch.bind(this),
+      });
     }
   }
 
   handleValue(value) {
     this.setState({
-      value,
+      value: value.value,
```

Comment:
Putting the function to fetch in state was a solution to having it re-render when changing the `sourcetype`, otherwise the options persisted. I'm changing it to use the key prop on `AsyncSelect` instead, removing the function from state.

Also I realized using `value.value` is major problem when setting a formula as the value is then just a string